File: src/main/java/org/lanternpowered/porygen/util/AbstractLine2.java
/*
 * This file is part of Porygen, licensed under the MIT License (MIT).
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the Software), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED AS IS, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package org.lanternpowered.porygen.util;

import static com.google.common.base.Preconditions.checkNotNull;

import com.google.common.base.MoreObjects;
import com.google.common.base.Objects;

abstract class AbstractLine2<P extends Comparable<P>> implements Line2<P> {

    final P start;
    final P end;

    AbstractLine2(P start, P end) {
        checkNotNull(start, "start");
        checkNotNull(end, "end");
        if (start.compareTo(end) < 0) {
            this.start = start;
            this.end = end;
        } else {
            this.start = end;
            this.end = start;
        }
    }

    @Override
    public P getStart() {
        return this.start;
    }

    @Override
    public P getEnd() {
        return this.end;
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("start", this.start)
                .add("end", this.end)
                .toString();
    }

    @Override
    public boolean equals(Object object) {
        if (object == null || object.getClass() != getClass()) {
            return false;
        }
        final AbstractLine2 other = (AbstractLine2) object;
        return other.start.equals(this.start) && other.end.equals(this.end);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(this.start, this.end);
    }
}

File: core/src/commonMain/kotlin/org/lanternpowered/porygen/math/geom/AbstractLine2.kt
/*
 * Porygen
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.porygen.math.geom

import org.lanternpowered.porygen.util.ToStringHelper

abstract class AbstractLine2<P : Comparable<P>> internal constructor() : Line2<P> {

  override fun toString(): String = ToStringHelper(this)
    .add("start", start)
    .add("end", end)
    .toString()

  override fun equals(other: Any?): Boolean {
    if (other == null || other::class != this::class)
      return false
    other as AbstractLine2<*>
    return other.start == start && other.end == end
  }

  override fun hashCode() = arrayOf<Any>(start, end).contentHashCode()
}
